salesagility/SuiteCRM
# FIX #3048 - Append and display messages by type.

## Description




This PR replaces outdated #3078
Feature of setting and displaying different types of color-coded messages for the user.

![captura](https://user-images.githubusercontent.com/939888/49826804-9601ec00-fd66-11e8-82f8-dafb27a1281f.png)


We had implemented this feature in SugarCRM 6.5 a while ago and now adapted the code for the current version of SuiteCRM.
Basically it allows to set messages by using static method SugarApplication::appendMessage, which is meant to be called from controller.php or any other place in the application. We also added the messages rendering to the header in SugarView.php and in the Smarty templates for each theme available, also setting up the code for the different CSS implementations currently used, in order to show color-coded errors with different styling classes.
As a second approach we also found useful to have a JS implementation for this functionality, so we have added that in sugar_3.js, using the functions displayMessage for the header area and displayMessagesAfterSelector for any other area you wish to show the messages on.
Also, this feature offers the possibility for implementing any new type of color-coded message you need. Now to accomplish that you only have to ensure that the corresponding CSS classes for your custom type exist and are available. For example:

```css
.myCustomType {
background-color: #700614;
color: #CCCCCC;
}
```
Also, as @samus-aran asked in old PR #3078  we consider to add code for accessible in the html div itself: role=alert


![codigo](https://user-images.githubusercontent.com/939888/49827434-0fe6a500-fd68-11e8-8113-1fe98f12a134.png)


## Motivation and Context


Please refer to issue: #3048

## How To Test This


You need to clean application cache. Also remember to repair compiled CSS stylesheets and minifyied JS cache files. Then run SugarApplicationTest.php with PHPUnit and/or call displayMessage function from JS console into to the browser. The type parameter currently has the following allowed values: info, error, alert, okay, working. When you use these remember to quote them as their expected data-type is string. See some examples below.

In javascript:

```
// Insert this directly into the browser console
displayMessage('working', 'working class hero');
```


In PHP:

```
// Put this in controller.php or any other PHP file in the view
SugarApplication::appendMessage('some info', 'info');
```

## Types of changes


### Final checklist
